### PR TITLE
#1-4 Task. TapeGhad

### PR DESCRIPTION
**Pull request by TapeGhad**
https://travis-ci.com/github/TapeGhad/js-assignments

### CODE REVIEW

**Comment by TapeGhad** [comment]
https://travis-ci.com/github/TapeGhad/js-assignments/builds/174226686

File: task/04-arrays-tasks.js [hunk 1]
```diff
@@ -116,7 +129,9 @@ function removeFalsyValues(arr) {
  *    [ 'a', 'b', 'c', 'd', 'e', 'f', 'g' ]  => [ 'A', 'B', 'C', 'D', 'E', 'F', 'G' ]
  */
 function getUpperCaseStrings(arr) {
-   throw new Error('Not implemented');
+   return arr.map(elem =>{
```
Comment (AlexMokin) [inline task/04-arrays-tasks.js]: array funcitons don't have to have a body like elem => {}, you can just use **elem => elem.toUpperCase()** for simple return statements

**Comment by AlexMokin** [comment]
General points:
1) Lot's of while-loop uses for a for-loop use case. 
2) lot's of missing ;
3) formatting
4) variable names

I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

File: task/04-arrays-tasks.js [hunk 2]
```diff
@@ -454,7 +511,11 @@ function getIdentityMatrix(n) {
  *     3, 3   => [ 3 ]
  */
 function getIntervalArray(start, end) {
-   throw new Error('Not implemented');
+   let newArr=new Array(end - start +1).fill(undefined), count=start-1;
```
Comment (AlexMokin) [inline task/04-arrays-tasks.js]: do not mix multiple variable declarations and the logic

**Comment by AlexMokin** [comment]
Please carefully review all comments. Not all issues are fixed.

Added a few new comments

I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by TapeGhad** [comment]
Want to add, that I changed as much as possible all points of your first comment. Moreover, this prase "Added a few new comments" from the second comment cornered me. I spend 10 minutes, checked js-assignment and all block and buttons at PR page, but couldn't find anything. May be I do something wrong or just missunderstood you

**Comment by AlexMokin** [comment]
Please carefully review all comments. Not all issues are fixed.
comments can't be found on the conversation tab. Alsos same comments are located in the code on the files changed tab

E.g. 
image


I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

File: task/01-strings-tasks.js [hunk 3]
```diff
@@ -233,13 +233,16 @@ function encodeToRot13(str) {
...
         if ((str.charCodeAt(count) >=78 & str.charCodeAt(count) <=90)| (str.charCodeAt(count) >=110 & str.charCodeAt(count)<=122))
+        {
             arr.push(str.charCodeAt(count)-13);
+        }
         else if ((str.charCodeAt(count) >=65 & str.charCodeAt(count) <=77)| (str.charCodeAt(count) >=97 & str.charCodeAt(count)<=109))
+        {
             arr.push(str.charCodeAt(count)+13);
+        }
         else arr.push(str.charCodeAt(count));
         count++;
```
Comment (TapeGhad) [inline task/01-strings-tasks.js]: Ок

File: task/01-strings-tasks.js [hunk 4]
```diff
@@ -221,7 +226,19 @@ function getRectangleString(width, height) {
...
+            arr.push(str.charCodeAt(count)-13);
+        }
+        else if ((str.charCodeAt(count) >=65 & str.charCodeAt(count) <=77)| (str.charCodeAt(count) >=97 & str.charCodeAt(count)<=109))
+        {
+            arr.push(str.charCodeAt(count)+13);
+        }
+        else arr.push(str.charCodeAt(count));
+    };
+    return String.fromCharCode(...arr);
 }
```
Comment (TapeGhad) [inline task/01-strings-tasks.js]: Here

**Comment by TapeGhad** [comment]
Hm. The problem is that I can't find any your comments, but can see mine after adding them. Something loke this
image

**Comment by TapeGhad** [comment]
Probably, after commit they disappear.

**Comment by TapeGhad** [comment]
Yes. Sorry for my commit after your screenshots. I didn't see your answer because I've been chaging while-loops on for-loops)

**Comment by AlexMokin** [comment]
Please carefully review all comments. Not all issues are fixed.

Also make sure local variable names follow google's naming convention for variable names
https://google.github.io/styleguide/jsguide.html#naming-local-variable-names. There are a lot variables in you code that are writen in snake_case, but all local variable names must follow camelCase

I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

File: task/04-arrays-tasks.js [hunk 5]
```diff
@@ -559,7 +637,21 @@ function getElementByIndexes(arr, indexes) {
...
+      newArr=newArr.concat(arr.slice(0, Math.floor(arr_length/2)));
+      return newArr;
+   }
+   else {
+      let newArr=[];
+      newArr= newArr.concat(arr.slice(arr_length/2));
+      newArr=newArr.concat(arr.slice(0, arr_length/2));
+      return newArr;
+    }
 }
```
Comment (TapeGhad) [inline task/04-arrays-tasks.js:662]: Oh

**Comment by TapeGhad** [comment]
May be I missunderstand something, but I can't see anyone comment in File changed block, instead of mine

**Comment by TapeGhad** [comment]
I suppose you should create some kind of review as I made above. Because, after checking information on the Internet, I can't find another way to leave and read comments without their displaying in Conversation block

File: task/01-strings-tasks.js [hunk 6]
```diff
@@ -69,7 +69,8 @@ function getStringFromTemplate(firstName, lastName) {
  *   'Hello, Chuck Norris!' => 'Chuck Norris'
  */
 function extractNameFromTemplate(value) {
-    throw new Error('Not implemented');
+    let name = value.slice(7);
+    return name.slice(0,-1);
```
Comment (AlexMokin) [inline task/01-strings-tasks.js]: simplify. Use only 1 slice

File: task/01-strings-tasks.js [hunk 7]
```diff
@@ -221,9 +226,20 @@ function getRectangleString(width, height) {
  *
  */
 function encodeToRot13(str) {
-    throw new Error('Not implemented');
+    let arr=[];
+    for (let count=0; count<str.length;count++){
+        if ((str.charCodeAt(count) >=78 & str.charCodeAt(count) <=90)| (str.charCodeAt(count) >=110 & str.charCodeAt(count)<=122))
```
Comment (AlexMokin) [inline task/01-strings-tasks.js]: do not use "magic numbers' https://en.wikipedia.org/wiki/Magic_number_(programming)

File: task/01-strings-tasks.js [hunk 8]
```diff
@@ -238,7 +254,8 @@ function encodeToRot13(str) {
  *   isString(new String('test')) => true
  */
 function isString(value) {
-    throw new Error('Not implemented');
+    if (typeof value ==="string" || value instanceof String) return true;
```
Comment (AlexMokin) [inline task/01-strings-tasks.js]: do not use pattern if(expression) return true else return false
(typeof value ==="string" || value instanceof String) is already a boolean value, just return it

File: task/02-numbers-tasks.js [hunk 9]
```diff
@@ -73,7 +75,7 @@ function getAverage(value1, value2) {
  *   (-5,0) (10,-10) => 18.027756377319946
  */
 function getDistanceBetweenPoints(x1, y1, x2, y2) {
-    throw new Error('Not implemented');
+    return Math.sqrt(Math.pow((x1-x2),2)+Math.pow((y1-y2),2));
```
Comment (AlexMokin) [inline task/02-numbers-tasks.js]: Math.hypot

File: task/02-numbers-tasks.js [hunk 10]
```diff
@@ -111,7 +113,7 @@ function getLinearEquationRoot(a, b) {
  *   (0,1) (1,2)     => 0
  */
 function getAngleBetweenVectors(x1, y1, x2, y2) {
-    throw new Error('Not implemented');
+    return Math.acos((x1*x2+y1*y2)/(Math.sqrt(Math.pow(x1,2)+Math.pow(y1,2)) * Math.sqrt(Math.sqrt(Math.pow(x2,2)+Math.pow(y2,2)))));
```
Comment (AlexMokin) [inline task/02-numbers-tasks.js]: simplify using Math.hypot

File: task/02-numbers-tasks.js [hunk 11]
```diff
@@ -160,7 +162,7 @@ function parseNumberFromString(value) {
  *   1,2,3   => 3.741657386773941
  */
 function getParallelipidedDiagonal(a,b,c) {
-    throw new Error('Not implemented');
+    return Math.sqrt(Math.pow(a,2) +Math.pow(b,2)+Math.pow(c,2));
```
Comment (AlexMokin) [inline task/02-numbers-tasks.js]: simplify using Math.hypot

File: task/02-numbers-tasks.js [hunk 12]
```diff
@@ -202,7 +205,15 @@ function roundToPowerOfTen(num, pow) {
  *   17 => true
  */
 function isPrime(n) {
-    throw new Error('Not implemented');
+    let check=true;
```
Comment (AlexMokin) [inline task/02-numbers-tasks.js]: use better variable name

File: task/02-numbers-tasks.js [hunk 13]
```diff
@@ -202,7 +205,15 @@ function roundToPowerOfTen(num, pow) {
...
+    for (let count=2; count<=n/2; count++){
+        if (n%count===0) 
+        {check = false;
+         break;
+        }
+    }
+    if (check===true) return true;
```
Comment (AlexMokin) [inline task/02-numbers-tasks.js]: just return bolean expression

File: task/02-numbers-tasks.js [hunk 14]
```diff
@@ -202,7 +205,15 @@ function roundToPowerOfTen(num, pow) {
  *   17 => true
  */
 function isPrime(n) {
-    throw new Error('Not implemented');
+    let check=true;
+    for (let count=2; count<=n/2; count++){
+        if (n%count===0) 
+        {check = false;
+         break;
```
Comment (AlexMokin) [inline task/02-numbers-tasks.js]: to interrup a loop you could just return false here

File: task/02-numbers-tasks.js [hunk 15]
```diff
@@ -202,7 +205,15 @@ function roundToPowerOfTen(num, pow) {
  *   17 => true
  */
 function isPrime(n) {
-    throw new Error('Not implemented');
+    let check=true;
+    for (let count=2; count<=n/2; count++){
+        if (n%count===0) 
+        {check = false;
```
Comment (AlexMokin) [inline task/02-numbers-tasks.js]: check = false; must be on a new line

File: task/03-date-tasks.js [hunk 16]
```diff
@@ -94,7 +107,8 @@ function timeSpanToString(startDate, endDate) {
  *    Date.UTC(2016,3,5,21, 0) => Math.PI/2
  */
 function angleBetweenClockHands(date) {
-    throw new Error('Not implemented');
+   let angleTime=Math.abs((date.getHours()*60+date.getMinutes()+date.getTimezoneOffset()-date.getMinutes()*12)/2)%360;
```
Comment (AlexMokin) [inline task/03-date-tasks.js]: this is unreadable. Try to reformat it, maybe create a few variables

File: task/04-arrays-tasks.js [hunk 17]
```diff
@@ -23,7 +23,8 @@
  *    [0, 1, 2, 3, 4, 5], 5    => 5
  */
 function findElement(arr, value) {
-   throw new Error('Not implemented');
+   if (arr.includes(value)) return arr.indexOf(value);
```
Comment (AlexMokin) [inline task/04-arrays-tasks.js]: check indexOf definition. If statement is redundant here

File: task/04-arrays-tasks.js [hunk 18]
```diff
@@ -321,7 +356,10 @@ function getPositivesCount(arr) {
  *   [ 'one','one','one','zero' ]     => [ 'zero','one','one','one' ]
  */
 function sortDigitNamesByNumericOrder(arr) {
-   throw new Error('Not implemented');
+   let newArr=[{"name": "zero","number":0},{"name": "one","number":1}, {"name": "two","number":2}, {"name": "three","number":3}, {"name": "four","number":4}, {"name": "five","number":5}, {"name": "six","number":6}, {"name": "seven","number":7}, {"name": "eight","number":8}, {"name": "nine","number":9}];
```
Comment (AlexMokin) [inline task/04-arrays-tasks.js]: - newArr is a bad variable name. 
- simplify use simple object {"zero":0, "one": 1 .....} or an array ['zero','one'...]

File: task/04-arrays-tasks.js [hunk 19]
```diff
@@ -536,7 +613,8 @@ function selectMany(arr, childrenSelector) {
  *   [[[ 1, 2, 3]]], [ 0, 0, 1 ]      => 2        (arr[0][0][1])
  */
 function getElementByIndexes(arr, indexes) {
-    throw new Error('Not implemented');
+   let place=indexes.map(elem=>{return `[${elem}]`}).join("");
```
Comment (AlexMokin) [inline task/04-arrays-tasks.js]: arrow functions doesn't have to have a body e.g.` let place = indexes.map(elem => '[${elem}]').join("");`

File: task/04-arrays-tasks.js [hunk 20]
```diff
@@ -536,7 +613,8 @@ function selectMany(arr, childrenSelector) {
...
+   return eval(`arr${place}`);
```
Comment (AlexMokin) [inline task/04-arrays-tasks.js]: never use eval

File: task/04-arrays-tasks.js [hunk 21]
```diff
@@ -437,7 +487,14 @@ function sortCitiesArray(arr) {
  *           [0,0,0,0,1]]   
  */
 function getIdentityMatrix(n) {
-   throw new Error('Not implemented');
+   let matrix= new Array(n).fill(undefined), countRow=0;
+   matrix = matrix.map(elem=>{
+      let newArr=new Array(n).fill(0);
```
Comment (AlexMokin) [inline task/04-arrays-tasks.js]: bad variable name

File: task/01-strings-tasks.js [hunk 22]
```diff
@@ -201,7 +202,11 @@ function extractEmails(str) {
  *
  */
 function getRectangleString(width, height) {
-    throw new Error('Not implemented');
+    let str="┌"+"─".repeat(width-2)+"┐"+'\n';
+    for (let count_height = height-2; count_height>0; count_height--){
```
Comment (AlexMokin) [inline task/01-strings-tasks.js]: check naming conventions for local variables - `count_height `

File: task/04-arrays-tasks.js [hunk 23]
```diff
@@ -437,7 +487,14 @@ function sortCitiesArray(arr) {
  *           [0,0,0,0,1]]   
  */
 function getIdentityMatrix(n) {
-   throw new Error('Not implemented');
+   let matrix= new Array(n).fill(undefined), countRow=0;
```
Comment (AlexMokin) [inline task/04-arrays-tasks.js]: do not mix multiple variable declarations and the logic

**Review by AlexMokin — COMMENTED** [review]
See the comments. Also check https://javascript.info/coding-style and fixe your code accordingly. 
Pay attention to how you should use spaces around operators

I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by AlexMokin** [comment]
Accepting but with issues. 
- Magic numbers issue not fixed.
- formatting issue
- mixing multiple variables declaration
- multiple use of the variables for different purposes 

image

**Pull request closed without merging**